File: android/nativebrik/src/main/java/com/nativebrik/sdk/data/track.kt
package com.nativebrik.sdk.data

import com.nativebrik.sdk.Config
import com.nativebrik.sdk.data.user.NativebrikUser
import com.nativebrik.sdk.data.user.formatISO8601
import com.nativebrik.sdk.data.user.getCurrentDate
import com.nativebrik.sdk.schema.ListDecoder
import com.nativebrik.sdk.schema.StringDecoder
import com.nativebrik.sdk.schema.TriggerEventNameDefs
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonObject
import okio.withLock
import java.time.ZonedDateTime
import java.util.Timer
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.fixedRateTimer

private val CRASH_RECORD_KEY = "CRASH_RECORD_KEY"

data class CrashRecord(
    val reason: String?,
    val callStacks: List<String>?
) {
    fun encode(): JsonObject {
        val callStacks = this.callStacks?.map {
            JsonPrimitive(it)
        } ?: emptyList()
        return JsonObject(mapOf(
            "reason" to JsonPrimitive(this.reason),
            "callStacks" to JsonArray(callStacks)
        ))
    }

    companion object {
        fun decode(element: JsonElement?): CrashRecord? {
            if (element == null) return null
            if (element !is JsonObject) return null

            return CrashRecord(
                reason = StringDecoder.decode(element.jsonObject["name"]),
                callStacks = ListDecoder.decode(element.jsonObject["callStacks"]) {
                    StringDecoder.decode(
                        it
                    )
                },
            )
        }
    }
}


internal data class TrackUserEvent(
    val name: String,
    val timestamp: ZonedDateTime = getCurrentDate(),
) {
    fun encode(): JsonObject {
        return JsonObject(mapOf(
            "typename" to JsonPrimitive("event"),
            "name" to JsonPrimitive(this.name),
            "timestamp" to JsonPrimitive(formatISO8601(this.timestamp)),
        ))
    }
}

internal data class TrackExperimentEvent(
    val experimentId: String,
    val variantId: String,
    val timestamp: ZonedDateTime = getCurrentDate(),
) {
    fun encode(): JsonObject {
        return JsonObject(mapOf(
            "typename" to JsonPrimitive("experiment"),
            "experimentId" to JsonPrimitive(this.experimentId),
            "variantId" to JsonPrimitive(this.variantId),
            "timestamp" to JsonPrimitive(formatISO8601(this.timestamp)),
        ))
    }
}

internal sealed class TrackEvent {
    class UserEvent(val event: TrackUserEvent) : TrackEvent()
    class ExperimentEvent(val event: TrackExperimentEvent) : TrackEvent()

    fun encode(): JsonObject {
        return when (this) {
            is UserEvent -> this.event.encode()
            is ExperimentEvent -> this.event.encode()
        }
    }
}

internal data class TrackRequest(
    val projectId: String,
    val userId: String,
    val events: List<TrackEvent>,
    val timestamp: ZonedDateTime = getCurrentDate(),
) {
    fun encode(): JsonObject {
        val events = this.events.map { it.encode() }
        return JsonObject(mapOf(
            "projectId" to JsonPrimitive(projectId),
            "userId" to JsonPrimitive(userId),
            "timestamp" to JsonPrimitive(formatISO8601(timestamp)),
            "events" to JsonArray(events)
        ))
    }
}

internal interface TrackRepository {
    fun trackExperimentEvent(event: TrackExperimentEvent)
    fun trackEvent(event: TrackUserEvent)

    fun record(throwable: Throwable)
}

internal class TrackRepositoryImpl: TrackRepository {
    private val queueLock: ReentrantLock = ReentrantLock()
    private val config: Config
    private val user: NativebrikUser
    private var timer: Timer? = null
    private val maxBatchSize: Int = 50
    private val maxQueueSize: Int = 300
    private var buffer: MutableList<TrackEvent> = mutableListOf()

    internal constructor(config: Config, user: NativebrikUser) {
        this.config = config
        this.user = user

        this.report()
    }

    override fun trackEvent(event: TrackUserEvent) {
        this.enqueue(TrackEvent.UserEvent(event))
    }

    override fun trackExperimentEvent(event: TrackExperimentEvent) {
        this.enqueue(TrackEvent.ExperimentEvent(event))
    }

    private fun enqueue(event: TrackEvent) {
        this.queueLock.withLock {
            if (this.timer == null) {
                val self = this
                CoroutineScope(Dispatchers.Main).launch {
                    self.timer?.cancel()
                    self.timer = fixedRateTimer(initialDelay = 0, period = 4000) {
                        CoroutineScope(Dispatchers.IO).launch {
                            self.sendAndFlush()
                        }
                    }
                }
            }
            if (this.buffer.size >= this.maxBatchSize) {
                val self = this
                CoroutineScope(Dispatchers.IO).launch {
                    self.sendAndFlush()
                }
            }
            this.buffer.add(event)
            if (buffer.size >= this.maxQueueSize) {
                this.buffer.drop(this.maxQueueSize - this.buffer.size)
            }
        }
    }

    private fun sendAndFlush() {
        val tempBuffer = this.buffer
        if (tempBuffer.isEmpty()) return
        this.buffer = mutableListOf()
        val request = TrackRequest(
            projectId = config.projectId,
            userId = user.id,
            events = tempBuffer
        )
        val body = Json.encodeToString(request.encode())
        this.timer?.cancel()
        this.timer = null
        postRequest(this.config.endpoint.track, body).onFailure {
            this.buffer.addAll(tempBuffer)
        }
    }

    private fun report() {
        val data = this.user.preferences?.getString(CRASH_RECORD_KEY, "") ?: ""
        if (data.isEmpty()) {
            return
        }
        this.user.preferences?.edit()?.remove(CRASH_RECORD_KEY)?.apply()

        val json = Json.decodeFromString<JsonElement>(data)
        val crashRecord = CrashRecord.decode(json) ?: return
        val causedByNativebrik = crashRecord.callStacks?.any { it.contains("com.nativebrik.sdk") } == true ||
                                 crashRecord.reason?.contains("com.nativebrik.sdk") == true

        this.buffer.add(TrackEvent.UserEvent(TrackUserEvent(
            name = TriggerEventNameDefs.N_ERROR_RECORD.name
        )))

        if (causedByNativebrik) {
            buffer.add(TrackEvent.UserEvent(TrackUserEvent(
                name = TriggerEventNameDefs.N_ERROR_IN_SDK_RECORD.name
            )))
        }

        val self = this
        CoroutineScope(Dispatchers.IO).launch {
            self.sendAndFlush()
        }
    }

    override fun record(throwable: Throwable) {
        val record = CrashRecord(
            reason = throwable.message,
            callStacks = throwable.stackTrace.map { it.toString() }
        )
        val data = Json.encodeToString(record.encode())
        this.user.preferences?.edit()?.putString(CRASH_RECORD_KEY, data)?.apply()
    }
}

File: ios/Nativebrik/Nativebrik/data/track.swift
//
//  remote-config.swift
//  Nativebrik
//
//  Created by Ryosuke Suzuki on 2024/03/07.
//

import Foundation

private let CRASH_RECORD_KEY: String = "NATIVEBRIK_CRASH_RECORD"

struct CrashRecord: Codable {
    var reason: String?
    var callStacks: [String]?
}

protocol TrackRepository2 {
    func trackExperimentEvent(_ event: TrackExperimentEvent)
    func trackEvent(_ event: TrackUserEvent)

    func record(_ exception: NSException)
}

struct TrackRequest: Encodable {
    var projectId: String
    var userId: String
    var timestamp: DateTime
    var events: [TrackEvent]
}

struct TrackEvent: Encodable {
    enum Typename: String, Encodable {
        case Event = "event"
        case Experiment = "experiment"
    }
    var typename: Typename
    var experimentId: String?
    var variantId: String?
    var name: String?
    var timestamp: DateTime
}

struct TrackUserEvent {
    var name: String
}

struct TrackExperimentEvent {
    var experimentId: String
    var variantId: String
}


class TrackRespositoryImpl: TrackRepository2 {
    private let maxQueueSize: Int
    private let maxBatchSize: Int
    private let config: Config
    private let user: NativebrikUser
    private let queueLock: NSLock
    private var timer: Timer?
    private var buffer: [TrackEvent]
    init(config: Config, user: NativebrikUser) {
        self.maxQueueSize = 300
        self.maxBatchSize = 50
        self.config = config
        self.user = user
        self.queueLock = NSLock()
        self.buffer = []
        self.timer = nil

        self.report()
    }

    deinit {
        self.timer?.invalidate()
    }

    func trackExperimentEvent(_ event: TrackExperimentEvent) {
        self.pushToQueue(TrackEvent(
            typename: .Experiment,
            experimentId: event.experimentId,
            variantId: event.variantId,
            timestamp: formatToISO8601(getCurrentDate())
        ))
    }

    func trackEvent(_ event: TrackUserEvent) {
        self.pushToQueue(TrackEvent(
            typename: .Event,
            name: event.name,
            timestamp: formatToISO8601(getCurrentDate())
        ))
    }

    private func pushToQueue(_ event: TrackEvent) {
        self.queueLock.lock()
        if self.timer == nil {
            // here, use async not sync. main.sync will break the app.
            DispatchQueue.main.async {
                self.timer?.invalidate()
                self.timer = Timer.scheduledTimer(withTimeInterval: 4.0, repeats: true, block: { _ in
                    Task(priority: .low) {
                        try await self.sendAndFlush()
                    }
                })
            }
        }

        if self.buffer.count >= self.maxBatchSize {
            Task(priority: .low) {
                try await self.sendAndFlush()
            }
        }
        self.buffer.append(event)
        if self.buffer.count >= self.maxQueueSize {
            self.buffer.removeFirst(self.maxQueueSize - self.buffer.count)
        }

        self.queueLock.unlock()
    }

    private func sendAndFlush() async throws {
        if self.buffer.count == 0 {
            return
        }
        let events = self.buffer
        self.buffer = []
        let trackRequest = TrackRequest(
            projectId: self.config.projectId,
            userId: self.user.id,
            timestamp: formatToISO8601(getCurrentDate()),
            events: events
        )

        do {
            let url = URL(string: config.trackUrl)!
            let jsonData = try JSONEncoder().encode(trackRequest)
            var request = URLRequest(url: url)
            request.httpMethod = "POST"
            request.httpBody = jsonData
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            let _ = try await nativebrikSession.data(for: request)

            self.timer?.invalidate()
            self.timer = nil
        } catch {
            self.buffer.append(contentsOf: events)
        }
    }

    private func report() {
        guard let data = self.user.userDB.data(forKey: CRASH_RECORD_KEY) else {
            return
        }
        do {
            self.user.userDB.removeObject(forKey: CRASH_RECORD_KEY)
            let crashRecord = try JSONDecoder().decode(CrashRecord.self, from: data)
            let causedByNativebrik = (
                crashRecord.callStacks?.contains(where: { callStack in
                return callStack.contains("Nativebrik")
            }) ?? false) || (crashRecord.reason?.contains("Nativebrik") ?? false)
            self.buffer.append(TrackEvent(
                typename: .Event,
                name: TriggerEventNameDefs.N_ERROR_RECORD.rawValue,
                timestamp: formatToISO8601(getCurrentDate())
            ))
            if causedByNativebrik {
                self.buffer.append(TrackEvent(
                    typename: .Event,
                    name: TriggerEventNameDefs.N_ERROR_IN_SDK_RECORD.rawValue,
                    timestamp: formatToISO8601(getCurrentDate())
                ))
            }

            Task(priority: .low) {
                try await self.sendAndFlush()
            }

        } catch {}
    }

    func record(_ exception: NSException) {
        let record = CrashRecord(
            reason: exception.reason,
            callStacks: exception.callStackSymbols
        )
        do {
            let json = try JSONEncoder().encode(record)
            self.user.userDB.set(json, forKey: CRASH_RECORD_KEY)
        } catch {}
    }
}
